pkg/buses/message_dispatcher_test.go

```diff
@@ -0,0 +1,380 @@
+					"x-request-id": {"altered-id"},
+					"knative-1":    {"new-knative-1-value"},
+					"ce-abc":       {"new-ce-abc-value"},
+				},
+				Body: "destination-response",
+			},
+		},
+	}
+	for n, tc := range testCases {
+		t.Run(n, func(t *testing.T) {
+			destHandler := &fakeHandler{
+				t:        t,
+				response: tc.fakeResponse,
+				requests: make([]requestValidation, 0),
+			}
+			destServer := httptest.NewServer(destHandler)
+			defer destServer.Close()
+			replyHandler := &fakeHandler{
+				t:        t,
+				response: tc.fakeResponse,
+				requests: make([]requestValidation, 0),
+			}
+			replyServer := httptest.NewServer(replyHandler)
+			defer replyServer.Close()
+
+			md := NewMessageDispatcher(zap.NewNop().Sugar())
+			err := md.DispatchMessage(tc.message,
+				getDomain(tc.sendToDestination, destServer.URL[7:]),
+				getDomain(tc.sendToReply, replyServer.URL[7:]),
+				DispatchDefaults{})
+			if tc.expectedErr != (err != nil) {
+				t.Errorf("Unexpected error from DispatchRequest. Expected %v. Actual: %v", tc.expectedErr, err)
+			}
+			if tc.expectedDestRequest != nil {
+				rv := destHandler.popRequest(t)
+				assertEquality(t, destServer.URL, *tc.expectedDestRequest, rv)
+			}
+			if tc.expectedReplyRequest != nil {
+				rv := replyHandler.popRequest(t)
+				assertEquality(t, replyServer.URL, *tc.expectedReplyRequest, rv)
+			}
+			if len(destHandler.requests) != 0 {
+				t.Errorf("Unexpected destination requests: %+v", destHandler.requests)
+			}
+			if len(replyHandler.requests) != 0 {
+				t.Errorf("Unexpected reply requests: %+v", replyHandler.requests)
+			}
+		})
+	}
+}
+
+func getDomain(shouldSend bool, domain string) string {
+	if shouldSend {
+		return domain
+	}
+	return ""
+}
+
+type requestValidation struct {
+	Host    string
+	Headers http.Header
+	Body    string
+}
+
+type fakeHandler struct {
+	t        *testing.T
+	response *http.Response
+	requests []requestValidation
+}
+
+func (f *fakeHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+
+	// Make a copy of the request.
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		f.t.Error("Failed to read the request body")
+	}
+	f.requests = append(f.requests, requestValidation{
+		Host:    r.Host,
+		Headers: r.Header,
+		Body:    string(body),
+	})
+
+	// Write the response.
+	if f.response != nil {
+		for h, vs := range f.response.Header {
+			for _, v := range vs {
+				w.Header().Add(h, v)
+			}
+		}
+		w.WriteHeader(f.response.StatusCode)
+		var buf bytes.Buffer
+		buf.ReadFrom(f.response.Body)
+		w.Write(buf.Bytes())
+	} else {
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte(""))
+	}
+}
+
+func (f *fakeHandler) popRequest(t *testing.T) requestValidation {
+	if len(f.requests) == 0 {
+		t.Error("Unable to pop request")
+	}
+	rv := f.requests[0]
+	f.requests = f.requests[1:]
+	return rv
+}
+
+func assertEquality(t *testing.T, replacementURL string, expected, actual requestValidation) {
+	expected.Host = replacementURL[7:]
```

Comment:
I presume this is stripping "http://"?

Reply:
This might be clearer as:
```golang
server, err := url.Parse(replacementURL)
if err != nil {
  t.Errorf("Bad replacement URL: %q", replacementURL)
}
expected.Host := server.Host
...
```

Reply:
Done, both here and in getDomain.